### PR TITLE
Fix #47, FQP initialization

### LINKED ISSUE #47: FQP initialization should take the mod of input coefficients
Full context: https://github.com/ethereum/eth2.0-specs/issues/508

### What is wrong?

For BN and BLS, FQP object should take the modulo of the input during initialization.

Test case:

```python
from py_ecc.optimized_bls12_381 import FQ, FQ2

x = FQ(-1)
y = FQ2([-1,-1])

print(x) # 4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559786
print(y) # (-1, -1)
```

This is used in  `modular_squareroot` part of Eth2.0 spec (https://github.com/ethereum/eth2.0-specs/blob/928f9772fa51ed2e251b99c4795097da976c4bde/specs/bls_signature.md#modular_squareroot) and Trinity (https://github.com/ethereum/trinity/blob/a1b0f058e7bc8e385c8dac3164c49098967fd5bb/eth2/_utils/bls.py#L63-L77)

### How can it be fixed

Take the mod of the inputs in FQP `__init__`

https://github.com/ethereum/py_ecc/blob/e9c42c76f1cc4f535dedec01b4c7ab2a59e46119/py_ecc/optimized_bls12_381/optimized_field_elements.py#L126-L133

Like it's done for FQ

https://github.com/ethereum/py_ecc/blob/e9c42c76f1cc4f535dedec01b4c7ab2a59e46119/py_ecc/optimized_bls12_381/optimized_field_elements.py#L25-L31

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### What was wrong?

#47 

### How was it fixed?

- Add tests of negative initialization parameters
- Make sure self.coeffs is moded if coeff are integers

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: py_ecc/optimized_bls12_381/optimized_field_elements.py [hunk 1]
```diff
@@ -239,7 +239,10 @@ def zero(cls):
 # The quadratic extension field
 class FQ2(FQP):
     def __init__(self, coeffs):
-        self.coeffs = tuple(coeffs)
+        if isinstance(coeffs[0], int):
+            self.coeffs = tuple(coeff % field_modulus for coeff in coeffs)
```
Comment (mratsim) [inline py_ecc/optimized_bls12_381/optimized_field_elements.py]: Wouldn't it be better to put that in the FQP class
and do:

`super(FQ2, self).__init__(coeffs, modulus_coeffs)`.

This way no need to duplicate the effort in FQ12 (and other extension fields).

Comment (Bhargavasomu) [reply py_ecc/optimized_bls12_381/optimized_field_elements.py]: I agree with @mratsim

Comment (ChihChengLiang) [reply py_ecc/optimized_bls12_381/optimized_field_elements.py]: Was about to do that, but realized that all __init__s are duplicated in `optimized_bls12_381`. A profile shows that using `super` in FQP would affect performance, so does the modulus check in this PR. Need to figure out a tolerable bound first. 

Master branch:
FQ2 1.123668404
FQ12 1.279635752

This PR:
FQ2 1.790521966
FQ12 2.642265235

Using Super in FQP:
FQ2 2.097532083
FQ12 3.0613951150000003

https://gist.github.com/ChihChengLiang/115ee83f1e5ce792db8ccf2392244455

Comment (ChihChengLiang) [reply py_ecc/optimized_bls12_381/optimized_field_elements.py]: Arrrr ... my bad, no need to check type. The input is always int.

Comment (ChihChengLiang) [reply py_ecc/optimized_bls12_381/optimized_field_elements.py]: Asked v about the performance, "get as much performance as we can" is desired.

Comment (ChihChengLiang) [reply py_ecc/optimized_bls12_381/optimized_field_elements.py]: Asked v for review. There are other codes pass FQs as coefficients to FQ2 and FQP, so `isinstance` check is required.

**Comment by ChihChengLiang** [comment]
Since there are changes in bn128, run a full test on py-evm and it passed. This is ready to merge. Thank you @mratsim @Bhargavasomu for the review.

**Comment by hwwhww** [comment]
@pipermerriam @carver 
Could you merge it and make new release of `py_ecc`? Thank you!

**Comment by carver** [comment]
Sorry for the delay!

**Comment by carver** [comment]
Tagged the next version, waiting on permissions to release it.
